src/pk/asn1/der/utf8/der_decode_utf8_string.c

```diff
@@ -53,21 +53,23 @@ int der_decode_utf8_string(const unsigned char *in,  unsigned long inlen,
 
    /* proceed to decode */
    for (y = 0; x < inlen; ) {
-      /* get first byte */
+      /* read first byte */
       tmp = in[x++];
 
-      /* count number of bytes */
+      /* count number of left-shifts needed to get zero in most-sig bit */
```

Comment:
IMO this comment makes it now even less clear, but this utf8 thing is also less than trivial... not to speak of well written documentation of source code ... ;)

What you described now is what is technically done, not what the purpose is.

The purpose of the for-loop is to count the number of most-significant bits which are set to 1.
This has to be done as the number of msb-1s tells us the number of bytes that are used to encode the utf8 symbol.

```
0 -> 1 byte
{2,3,4} -> {2,3,4} bytes
```

Reply:
I didn't want to try and explain too much in this first comment.  I just wanted to explain the for-loop.  You're right that we are counting the number of ms-bits equal to one, but the way we count that is by left-shifting until the ms-bit is zero.

Note that there is a comment below that explains how `z` gives the number of additional bytes to read.

I can add an additional comment above the line `tmp = in[x++]` explaining the utf-8 decoding strategy, if that would help.